keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y/README.md

```diff
@@ -0,0 +1,606 @@
+and not the version of the kubelet on the host where `kubeadm init` is called.
+
+Later, during `join` operations the version `x.y` is determined by looking up the control plane version
+of the cluster, via the `kubernetesVersion` field of the `ClusterConfiguration` object stored
+in the `kube-system/kubeadm-config` ConfigMap.
+
+During `upgrade` a new `kubelet-config-x.y+1` ConfigMap and RBAC rules are written where `y+1` is the target
+MINOR upgrade version. The old `kubelet-config-x.y` are not cleaned up.
+
+Encoding the version in the API object names is redundant and creates unnecessary complications for both
+kubeadm maintainers and users that wish to modify the object in the cluster.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Simplify the name of the kubelet related ConfigMap and RBAC rules that kubeadm manages to exclude
+encoded version information in the API object names.
+- Remove confusion around the version that is encoded in the name - control plane vs kubelet version?
+- Simplify user access to the ConfigMap, allowing them to always rely on `kube-system/kubelet-config`
+as the default location.
+- Simplify the kubeadm logic to not manage the `x.y` version during PUT/GET operations.
+- Gradually apply the change to allow users to adapt using a feature gate.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Apply the change within a single release of kubeadm, breaking the users.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal is to declare the default ConfigMap and RBAC object format of `kubelet-config-x.y`
+deprecated but support it for a GA deprecation period of 1 year (3 kubeadm releases). In parallel,
+introduce objects named `kubelet-config` (without `x.y`) that will become the source of truth.
+
+The switch will be controlled by a new kubeadm specific feature gate (not a core one) called
+`UnversionedKubeletConfigMap`. Using a feature gate for similar changes is common in changes
+in core Kubernetes components and is familiar mechanism to users.
+
+Note that depending on feedback from users the N+x release schedule might be adjusted - e.g.
+extending or shortening the Alpha or Beta periods.
+
+#### First release N (when the deprecation happens)
+
+- The feature gate is added as Alpha and disabled by default.
+- Users can continue using the old naming format or opt-in into the new naming format
+by enabling the feature gate.
```

Comment:
Instead of having a feature gate, have we considered writing both maps at the same time and switch the functionality over time?

Reply:
this was my first approach on the topic, but i estimated that it would require a hard switch at some point. i have covered it briefly in Alternatives (could expand more there).

we could start writing both ConfigMaps for a few releases and then stop writing the old format after the deprecation, but it seems there is a problem with reading. kubeadm as a client would have to use the old format for a few releases and then suddenly switch to the new format. that would be a hard switch withing one release without the option for users to opt-out. a feature gate during the Beta period would allow them the opt-out.

for reading, we could implement a soft-switch of sorts (for e.g. one release) where the contents of both ConfigMaps are combined, considering both of them as the source of truth, but that feels very messy - e.g. trying to guess what field value the user really wants to use.

Reply:
> we could start writing both ConfigMaps for a few releases and then stop writing the old format after the deprecation, but it seems there is a problem with reading. kubeadm as a client would have to use the old format for a few releases and then suddenly switch to the new format. that would be a hard switch within one release without the option for users to opt-out. a feature gate during the Beta period would allow them the opt-out.

I agree that a hard switch can be overwhelming for users at a time, although with proper communication and by spreading the change in a 3 or more releases seems like a good way to migrate. I was thinking of something like:

- Write both configmaps in a release (let's say 1.23) + give a release note to users that the older configmaps are going to be non-functional in a future release.
- Start ignoring the older format in 1.24 (making it non-functional) and alerting users through release notes and required changes.
- Removing and cleaning up the older code.

The reason I'm suggesting this approach is to avoid maintaining two code-paths for longer period of times with a feature gate, the solution proposed above seems more a migration than a new feature, although I'm fine either way.

Having a clear cut-date for the migration also helps kubeadm users that have to support multiple versions. Cluster API for example can start detecting the change by version, rather than by version + feature gate and migrating clusters on the fly. I'd assume other users might want to do something similar.

Reply:
> I agree that a hard switch can be overwhelming for users at a time, although with proper communication and by spreading the change in a 3 or more releases seems like a good way to migrate. I was thinking of something like:

we've decided on a hard switch for the deprecation of the `master` label and taint - without a feature gate over a period of time because a feature gate made it more difficult:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint#using-a-feature-gate
but i still have some concerns that between some of the stages we are going to break users badly.

as the main problem is that a good number of users simply skip reading the release notes.

> The reason I'm suggesting this approach is to avoid maintaining two code-paths for longer period of times with a feature gate, 

the two code paths would be mostly trivial with a ClusterConfiguration.FeatureGates read controlling the formatting of a string.

> the solution proposed above seems more a migration than a new feature, although I'm fine either way.

i'm also fine either way, and would like to hear more opinions before Sept 9th.

> Having a clear cut-date for the migration also helps kubeadm users that have to support multiple versions. Cluster API for example can start detecting the change by version, rather than by version + feature gate and migrating clusters on the fly. I'd assume other users might want to do something similar.

with or without the feature gate idea CAPI would have to adapt for it's upgrade code to populate the missing API objects normally maintained by `kubeadm upgrade apply`. a version check only (==1.24) would be sufficient for CAPI to decide if it has to create a copy of the existing objects into objects with the new name format. i don't think it has to check the state of the feature gate.

AFAIK, CAPI currently has the feature gates of the ClusterConfiguration immutable, which effectively means CAPI users cannot use them:
https://github.com/kubernetes-sigs/cluster-api/issues/4180

if they can use them and if we go the FG route, this means that CAPI should adapt to the "copy logic" above in 1.23 (Alpha, not 1.24 / Beta).

Reply:
I'm +1 for the feature gate given that this allows users some flexibility in deciding when to switch (and given that the implementation effort seems reasonable).
From a CAPI PoV, I will suggest we wait for this change to become beta active but default; this will allows us to just work by version (but with an escape patch via the feature flag if we decide to wait a little bit longer for any reason).